compat/mingw.c

```diff
@@ -426,6 +426,43 @@ static wchar_t *normalize_ntpath(wchar_t *wbuf)
 	return wbuf;
 }
 
+/* rename and unlink a file. This is needed for the ability to recreate a file
+ * with the same name, for example on checkout which unlinks and recreates the file */
+static int rename_and_unlink(const wchar_t *pathname, int (*unlinker)(const wchar_t *))
+{
+	wchar_t temppath[MAX_LONG_PATH];
+	*temppath = 0;
+	/* leave some space for the .unlink_XXXXXX extension */
+	wcsncat(temppath, pathname, sizeof(temppath) - 15);
+	wcsncat(temppath, L".unlink_XXXXXX", sizeof(temppath));
+	_wmktemp(temppath);
+	if (MoveFileExW(pathname, temppath, 0)) {
+		const int res = unlinker(temppath);
+		if (res) {
+			const DWORD err = GetLastError();
```

Comment:
For the error handling I suggest that we do the rename in `mingw_unlink`, before the `do_unlink` retry loop. If the rename is also prone to failures and needs to be retried, then add a separate retry loop for the rename. Return with an error if the rename fails or if we give up retrying.

Then, do the unlink on the renamed file normally, without worrying about renaming.

Reply:
Done